src/worker.cc

```diff
@@ -0,0 +1,699 @@
+
+  owner_env->AddSubWorkerContext(this);
+
+  CHECK_EQ(uv_mutex_init(termination_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(api_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(initialization_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(to_owner_messages_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(to_worker_messages_mutex()), 0);
+}
+
+void WorkerContext::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope scope(env->isolate());
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsObject());
+
+  Local<String> entry_module_path = args[0].As<String>();
+  Local<Object> options = args[1].As<Object>();
+
+#define GET_OPTION(var_name, option_name, Type)                               \
+  Local<Type> var_name;                                                       \
+  {                                                                           \
+    Local<Value> val = options->Get(FIXED_ONE_BYTE_STRING(env->isolate(),     \
+                                                          #option_name));     \
+    var_name = val.As<Type>();                                                \
+  }                                                                           \
+
+  GET_OPTION(keep_alive, keepAlive, Value)
+  GET_OPTION(user_data, userData, String)
+  GET_OPTION(eval, eval, Value)
+#undef GET_OPTION
+
+
+
+  uv_loop_t* worker_event_loop = static_cast<uv_loop_t*>(
+      malloc(sizeof(uv_loop_t)));
+  int err = uv_loop_init(worker_event_loop);
+  if (err != 0) {
+    free(worker_event_loop);
+    return env->ThrowUVException(err, "uv_loop_init");
+  }
+
+  // Heap allocated cause this stack frame can be gone by the time the worker
+  // needs them.
+  int argc = 2;
+  char** argv = new char*[argc];
+  // TODO(petkaantonov) should use correct executable name here
+  argv[0] = const_cast<char*>("node");
+  argv[1] = ToUtf8Value(env->isolate(), entry_module_path);
+  // TODO(petkaantonov) should use the process's real exec args
+  int exec_argc = 0;
+  char** exec_argv = new char*[exec_argc];
+
+  WorkerContext* context =
+      new WorkerContext(env,
+                        args.This(),
+                        worker_event_loop,
+                        argc,
+                        const_cast<const char**>(argv),
+                        exec_argc,
+                        const_cast<const char**>(exec_argv),
+                        keep_alive->BooleanValue(),
+                        ToUtf8Value(env->isolate(), user_data),
+                        eval->BooleanValue());
+
+  err = uv_thread_create(context->worker_thread(), RunWorkerThread, context);
+
+  if (err != 0) {
+    context->worker_notifications()->Dispose();
+    context->Dispose();
+    CHECK_EQ(uv_loop_close(worker_event_loop), 0);
+    free(worker_event_loop);
+    return env->ThrowUVException(err, "uv_thread_create");
+  }
+}
+
+void WorkerContext::PostMessageToOwner(WorkerMessage* message) {
+  if (termination_kind() != NONE) {
```

Comment:
I'll try that, but I don't see how it could help. `CanCallIntoJs()` in a worker thread is equivalent to an unlocked check of `termination_kind() != ABRUPT`. Since that check is a data race with the owner thread, not only could `termination_kind_` become equal to `ABRUPT` immediately after we call `CanCallIntoJs()`, but `termination_kind_` could *already* have been set to `ABRUPT` on another core.

Anyway, as I said, I'll give it a try. Thanks for your help.